pkg/conference/start.go

```diff
@@ -47,23 +47,18 @@ func StartConference(
 		matrixWorker:      newMatrixWorker(signaling),
 		tracker:           *participant.NewParticipantTracker(),
 		streamsMetadata:   make(event.CallSDPStreamMetadata),
-		endNotifier:       conferenceEndNotifier,
-		peerMessages:      make(chan common.Message[participant.ID, peer.MessageContent], common.UnboundedChannelSize),
-		matrixMessages:    receiver,
+		peerMessages:      make(chan channel.Message[participant.ID, peer.MessageContent], 100),
+		matrixEvents:      matrixEvents,
+		conferenceDone:    done,
 	}
 
 	participantID := participant.ID{UserID: userID, DeviceID: inviteEvent.DeviceID, CallID: inviteEvent.CallID}
 	if err := conference.onNewParticipant(participantID, inviteEvent); err != nil {
-		return nil, err
+		return nil, nil
 	}
 
 	// Start conference "main loop".
 	go conference.processMessages()
```

Comment:
Maybe  if processMessages has its own done channel that it closes, then we can wait here for the processMessages done and clean up if necessary.

```go
go func() {
    defer close(done) // close conference after processMessages is done
    pmDone := conference.processMessages()
    <- pmDone
    // ... do some clean ups, currently not needed
}()   
```
I thought only because the `conference done` is distributed in the sinks. 
And it would have the advantage that we don't have to bind `done` it to the conference structure.
What do you think?

Reply:
I think your point about storing `done` inside the conference is valid. This could be solved if I pass `done` to the `processMessages` (so that we won't need to store it in the conference and rename it to `signalDone` instead of `done` so that it does not give a notion as if it's something that **we** need to listen to).

The rest I think is identical (we start  go-routine for processMessages that then cleans up things), though we don't need to wait on `done` inside `StartConference()`, since that `done` channel is returned for the upstream jobs, so that they know when the conference is closed.

I.e. we return the `done`, so that the caller which sends messages to the conference knows when to stop producing new values (when to stop sending values to the conference).

That way when the sender sends a matrix message to the conference, the sender's code looks like this:

```go
select {
    case conferenceSink <- message:
        // message sent to the conference, cool
    case <-conferenceDone:
        // conference closed, ok, so we don't send messages to it anymore
}
```

Reply:
I agree. At this time we are not running in a panic because the conference message channel will never be closed. With the clean up we could first close the sinks and then close the message channel from conference.

Reply:
> At this time we are not running in a panic because the conference message channel will never be closed.

Sorry, I did not quite get it. The channel between the router and a conference gets closed by the sender (router) once the sender spots that the `Conference` closed the `done` channel. The `done` channel is closed by the conference when the conference ends to inform the upstream stage (router) that no more new messages are expected. When the router spots that, the router closes the message channel that it used to send messages to the conference and removes the conference from the list.

Reply:
I meant this select construct [select to send to conference](https://github.com/matrix-org/waterfall/blob/4ae95a8fa7494093344f443be72bf1224fe3d277/pkg/channel/sink.go#L53)
could lead to a panic when this [conference.peerMessages](https://github.com/matrix-org/waterfall/blob/4ae95a8fa7494093344f443be72bf1224fe3d277/pkg/conference/start.go#L50) channel would be closed. Because every Sink writes to this channel. 

Here in [NewSink](https://github.com/matrix-org/waterfall/blob/4ae95a8fa7494093344f443be72bf1224fe3d277/pkg/conference/matrix_message_processing.go#L50) the channel `conference.peerMessages` is handed over. I couldn't find a place where we close `conference.peerMessages`. 

Should we decide to close `conference.peerMessages` then here would be. a good place, after all sinks and processMessages have been closed.

Reply:
Ah, I see, you mean a different place. Yes, this is a single leftover that is still not refactored (I mentioned it in the description), i.e. ideally that part should be rewritten as a fan-in thing where we have a single sender for each peer and then a `merge` function. Though we would need to use `Select` from the `reflect` since the number of peers changes over time.